src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.Caching.cs

```diff
@@ -163,146 +162,124 @@ public void Clear()
 
         /// <summary>
         /// Defines a cache of CachingContexts; instead of using a ConditionalWeakTable which can be slow to traverse
-        /// this approach uses a concurrent dictionary pointing to weak references of <see cref="CachingContext"/>.
-        /// Relevant caching contexts are looked up using the equality comparison defined by <see cref="EqualityComparer"/>.
+        /// this approach uses a fixed-size array of weak references of <see cref="CachingContext"/> that can be looked up lock-free.
+        /// Relevant caching contexts are looked up by linear traversal using the equality comparison defined by <see cref="EqualityComparer"/>.
         /// </summary>
         internal static class TrackedCachingContexts
         {
             private const int MaxTrackedContexts = 64;
-            private static readonly ConcurrentDictionary<JsonSerializerOptions, WeakReference<CachingContext>> s_cache =
-                new(concurrencyLevel: 1, capacity: MaxTrackedContexts, new EqualityComparer());
+            private static readonly WeakReference<CachingContext>?[] s_trackedContexts = new WeakReference<CachingContext>[MaxTrackedContexts];
```

Comment:
Entries in the cache could end up being removed even on a gen0 GC if there's no rooted JsonSerializerOptions currently referencing that CachingContext.  How likely is that?

Reply:
It is likely, although this is also true for the current implementation which also stores the cache inside a WeakReference.

Reply:
Yes, this cache works nice in micro-benchmarks and small apps. It is unlikely to be effective in larger apps that do number of different things where the GC is going to be likely to clear the weak references too often.

Reply:
We could replace the WeakReferences altogether with a manual eviction mechanism with an explicit TTL, although this will most likely involve setting up a `Timer` in the background. I think we discussed doing that last year but I couldn't find any prior art of other libraries doing anything similar.

Reply:
The timer just comes with different set of problems. I do not think that it makes sense to magically fix the problem for everybody.

Instead, we should spend energy on making sure that people do their own caching correctly and avoid this problem all together.

Reply:
We should have an analyzer that flags these kinds of cases, in particular constructing a new JsonSerializerOptions for each Serialize call, especially if its only populated with constants / always the same way.

We still need a cache, though, even if it doesn't help all cases.  The cliff is too steep otherwise.

Reply:
Related: https://github.com/dotnet/runtime/issues/65396

Reply:
> I do not think that it makes sense to try to magically fix the problem for everybody.

It certainly cannot fix the problem for all use cases. The equality comparison used to determine configuration equivalence is pretty conservative. For example, it only compares custom converters up to reference equality, which is unlikely to happen in cases where custom converters are being configured.